src/main/java/muramasa/antimatter/cover/CoverInstance.java

```diff
@@ -0,0 +1,129 @@
+package muramasa.antimatter.cover;
+
+import muramasa.antimatter.Data;
+import muramasa.antimatter.gui.GuiData;
+import muramasa.antimatter.machine.event.IMachineEvent;
+import muramasa.antimatter.tile.TileEntityMachine;
+import muramasa.antimatter.tool.AntimatterToolType;
+import net.minecraft.entity.item.ItemEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.PlayerInventory;
+import net.minecraft.inventory.container.Container;
+import net.minecraft.inventory.container.INamedContainerProvider;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.Direction;
+import net.minecraft.util.Hand;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.StringTextComponent;
+import net.minecraft.util.text.TranslationTextComponent;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Objects;
+import java.util.Optional;
+
+public class CoverInstance implements INamedContainerProvider {
+    private Cover cover;
+    private CompoundNBT nbt;
+    private TileEntity tile;
+
+    public CoverInstance(Cover cover, TileEntity tile) {
+        this.cover = Objects.requireNonNull(cover);
+        this.tile = tile;
+        this.nbt = new CompoundNBT();
+    }
+    //This allows you to instantiate a non-stateful cover, like COVER_EMPTY.
+    //Using state with this is a runtime error.
+    public CoverInstance(Cover cover) {
+        this.nbt = new CompoundNBT();
+        this.cover = cover;
+    }
+
+    //Automatically calls onPlace.
+    public CoverInstance(Cover cover, TileEntity tile, Direction side) {
+        this(cover,tile);
+        onPlace(side);
+    }
+
+    public void serialize(CompoundNBT nbt) {
+        nbt.putString("ID",cover.getId());
+    }
+
+    public boolean isEqual(Cover cover) {
+        return this.cover.getId().equals(cover.getId());
+    }
+
+    public boolean isEqual(CoverInstance cover) {
+        return this.cover.getId().equals(cover.cover.getId());
+    }
+
+    public String getId() {
+        return this.cover.getId();
+    }
+
+    public boolean isEmpty() {
+        return cover == Data.COVERNONE;
+    }
+
+    //Gets the backing cover.
+    //Because getCover().getCover() looks stupid
+    public Cover backing() {
```

Comment:
backing is weird tho

Reply:
yeah but i always wrote getCover() to get the instance then getCover() to get the singleton, sounds stoopid

Reply:
I am advising renaming to the asCover() because backing() even more confusing

Reply:
can change to asCover

Reply:
Another idea it revert backing to getCover() and replace  
> public CoverInstance getCover(Direction side)
by the 
> public CoverInstance getCoverInstance(Direction side) or [getInstance]
which even should be better